docs/proposals/manifest-hydrator.md

```diff
@@ -0,0 +1,308 @@
+
+The actual push operation should be delegated to some system outside the application controller. Communication may occur via some shared DB (maybe Redis) or via network, e.g. gRPC.
+
+To understand how this would work for a simple dev/test/prod setup with two regions, consider this example:
+
+```yaml
+### DEV APPS ###
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: dev-west
+spec:
+  sourceHydrator:
+    drySources:
+    - repoURL: https://github.com/argoproj/argocd-example-apps
+      targetRevision: main
+      path: environments/dev/west
+    writeTo:
+      targetBranch: environments/dev
+      path: west
+---
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: dev-east
+spec:
+  sourceHydrator:
+    drySources:
+      - repoURL: https://github.com/argoproj/argocd-example-apps
+        targetRevision: main
+        path: environments/dev/east
+    writeTo:
+      targetBranch: environments/dev
+      path: east
+---
+### TEST APPS ###
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: test-west
+spec:
+  sourceHydrator:
+    drySources:
+      - repoURL: https://github.com/argoproj/argocd-example-apps
+        targetRevision: main
+        path: environments/test/west
+    writeTo:
+      targetBranch: environments/test
+      path: west
+---
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: test-east
+spec:
+  sourceHydrator:
+    drySources:
+      - repoURL: https://github.com/argoproj/argocd-example-apps
+        targetRevision: main
+        path: environments/test/east
+    writeTo:
+      targetBranch: environments/prod
+      path: east
+---
+### PROD APPS ###
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: prod-west
+spec:
+  sourceHydrator:
+    drySources:
+      - repoURL: https://github.com/argoproj/argocd-example-apps
+        targetRevision: main
+        path: environments/prod/west
+    writeTo:
+      targetBranch: environments/prod
+      path: west
+---
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: prod-east
+spec:
+  sourceHydrator:
+    drySources:
+      - repoURL: https://github.com/argoproj/argocd-example-apps
+        targetRevision: main
+        path: environments/prod/east
+    writeTo:
+      targetBranch: environments/prod
+      path: east
+---
+```
+
+Each commit to the dry branch will result in a commit to up to three branches. Each commit to an environment branch will contain changes for west, east, or both (depending on which is affected). Changes originating from a single dry commit are always grouped into a single hydrated commit.
```

Comment:
When implementing the rendered manifest pattern, we have found that the diff of the rendered manifests is quite noisy and our users have found that their intended change is lost in the noise. As a result, they often don't review the rendered manifests at all. In order to solve this problem, we publish a shortened diff in the comment of the PR to the interface repository. That diff intentionally omits certain boilerplate changes, in particular automatically generated metadata (for example, I link back to the PR that the automation adds as an annotation, config map name change, etc.). Users still have access to the full diff, however, available under a link automatically added to the PR. We have found that reviewing the dry change + shortened hydrated change is a better user experience than reviewing the entire noisy change to the rendered manifests.

Reply:
Interesting... do you know of a general way to simplify diffs? Or does it have to be org-specific?

I generally believe that writing org-specific CRDs+controllers is the best way to hide unnecessary detail from devs. But in practice I know that won't always be practical.

Reply:
+1 for the configmap workaround for rollouts. But you could handle that with a 
```
argocd.argoproj.io/sync-options: PruneLast=true
```

On the configmap no ?

Reply:
I feel like 1-4 would be great for basically any org. I'm not sure whether those should be part of this proposal... I tend towards saying "probably not."